File: attocode/src/integrations/budget/resources.ts
/**
 * Resource Monitor Integration
 *
 * Provides lightweight resource usage tracking and limits for agent operations.
 * Integrates with the economics system to check both token budget AND system resources.
 * Adapted from tricks/resource-monitor.ts.
 *
 * Usage:
 *   const resources = createResourceManager({ maxMemoryMB: 512 });
 *   const check = resources.check();
 *   if (check.status === 'exceeded') { ... }
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Resource limits configuration.
 */
export interface ResourceLimitsConfig {
  /** Enable/disable resource monitoring */
  enabled?: boolean;
  /** Max memory in MB */
  maxMemoryMB?: number;
  /** Max CPU time in seconds */
  maxCpuTimeSec?: number;
  /** Max concurrent operations */
  maxConcurrentOps?: number;
  /** Warning threshold (0-1) - emit warning when usage exceeds this */
  warnThreshold?: number;
  /** Critical threshold (0-1) - slow down operations when exceeded */
  criticalThreshold?: number;
}

/**
 * Current resource usage.
 */
export interface ResourceUsage {
  memoryMB: number;
  memoryPercent: number;
  cpuTimeSec: number;
  cpuPercent: number;
  concurrentOps: number;
  opsPercent: number;
  timestamp: Date;
}

/**
 * Resource status levels.
 */
export type ResourceStatus = 'healthy' | 'warning' | 'critical' | 'exceeded';

/**
 * Result of a resource check.
 */
export interface ResourceCheck {
  status: ResourceStatus;
  usage: ResourceUsage;
  canContinue: boolean;
  message?: string;
  recommendation?: 'continue' | 'slow_down' | 'stop';
}

/**
 * Resource event types.
 */
export type ResourceEvent =
  | { type: 'resource.warning'; metric: string; value: number; threshold: number }
  | { type: 'resource.critical'; metric: string; value: number; threshold: number }
  | { type: 'resource.exceeded'; metric: string; value: number; limit: number }
  | { type: 'resource.recovered'; metric: string; previousStatus: ResourceStatus };

export type ResourceEventListener = (event: ResourceEvent) => void;

// =============================================================================
// RESOURCE MANAGER
// =============================================================================

/**
 * Manages resource monitoring and limits for the agent.
 */
export class ResourceManager {
  private config: Required<ResourceLimitsConfig>;
  private startTime: number;
  private concurrentOps: number = 0;
  private previousStatus: ResourceStatus = 'healthy';
  private eventListeners: Set<ResourceEventListener> = new Set();

  constructor(config: ResourceLimitsConfig = {}) {
    this.config = {
      enabled: config.enabled ?? true,
      maxMemoryMB: config.maxMemoryMB ?? 512,
      maxCpuTimeSec: config.maxCpuTimeSec ?? 300, // 5 minutes
      maxConcurrentOps: config.maxConcurrentOps ?? 10,
      warnThreshold: config.warnThreshold ?? 0.7,
      criticalThreshold: config.criticalThreshold ?? 0.9,
    };
    this.startTime = Date.now();
  }

  /**
   * Get current resource usage.
   */
  getUsage(): ResourceUsage {
    const mem = process.memoryUsage();
    const memoryMB = mem.heapUsed / 1024 / 1024;
    const cpuTimeSec = (Date.now() - this.startTime) / 1000;

    return {
      memoryMB,
      memoryPercent: memoryMB / this.config.maxMemoryMB,
      cpuTimeSec,
      cpuPercent: cpuTimeSec / this.config.maxCpuTimeSec,
      concurrentOps: this.concurrentOps,
      opsPercent: this.concurrentOps / this.config.maxConcurrentOps,
      timestamp: new Date(),
    };
  }

  /**
   * Check resource status and determine if operations can continue.
   */
  check(): ResourceCheck {
    if (!this.config.enabled) {
      return {
        status: 'healthy',
        usage: this.getUsage(),
        canContinue: true,
        recommendation: 'continue',
      };
    }

    const usage = this.getUsage();
    let status: ResourceStatus = 'healthy';
    let message: string | undefined;
    let canContinue = true;
    let recommendation: 'continue' | 'slow_down' | 'stop' = 'continue';

    // Check memory
    if (usage.memoryPercent >= 1) {
      status = 'exceeded';
      message = `Memory limit exceeded (${usage.memoryMB.toFixed(1)}MB / ${this.config.maxMemoryMB}MB)`;
      canContinue = false;
      recommendation = 'stop';
      this.emit({
        type: 'resource.exceeded',
        metric: 'memory',
        value: usage.memoryMB,
        limit: this.config.maxMemoryMB,
      });
    } else if (usage.memoryPercent >= this.config.criticalThreshold) {
      status = 'critical';
      message = `Memory critically high (${(usage.memoryPercent * 100).toFixed(1)}%)`;
      recommendation = 'slow_down';
      this.emit({
        type: 'resource.critical',
        metric: 'memory',
        value: usage.memoryPercent,
        threshold: this.config.criticalThreshold,
      });
    } else if (usage.memoryPercent >= this.config.warnThreshold) {
      status = 'warning';
      message = `Memory usage elevated (${(usage.memoryPercent * 100).toFixed(1)}%)`;
      this.emit({
        type: 'resource.warning',
        metric: 'memory',
        value: usage.memoryPercent,
        threshold: this.config.warnThreshold,
      });
    }

    // Check CPU time (only override if worse)
    if (usage.cpuPercent >= 1 && status !== 'exceeded') {
      status = 'exceeded';
      message = `CPU time limit exceeded (${usage.cpuTimeSec.toFixed(1)}s / ${this.config.maxCpuTimeSec}s)`;
      canContinue = false;
      recommendation = 'stop';
      this.emit({
        type: 'resource.exceeded',
        metric: 'cpuTime',
        value: usage.cpuTimeSec,
        limit: this.config.maxCpuTimeSec,
      });
    } else if (usage.cpuPercent >= this.config.criticalThreshold && status === 'healthy') {
      status = 'critical';
      message = `Approaching CPU time limit (${(usage.cpuPercent * 100).toFixed(1)}%)`;
      recommendation = 'slow_down';
      this.emit({
        type: 'resource.critical',
        metric: 'cpuTime',
        value: usage.cpuPercent,
        threshold: this.config.criticalThreshold,
      });
    } else if (usage.cpuPercent >= this.config.warnThreshold && status === 'healthy') {
      status = 'warning';
      message = `CPU time usage elevated (${(usage.cpuPercent * 100).toFixed(1)}%)`;
      this.emit({
        type: 'resource.warning',
        metric: 'cpuTime',
        value: usage.cpuPercent,
        threshold: this.config.warnThreshold,
      });
    }

    // Check concurrent operations
    if (usage.opsPercent >= 1 && status !== 'exceeded') {
      status = 'exceeded';
      message = `Max concurrent operations reached (${usage.concurrentOps}/${this.config.maxConcurrentOps})`;
      canContinue = false;
      recommendation = 'stop';
      this.emit({
        type: 'resource.exceeded',
        metric: 'concurrentOps',
        value: usage.concurrentOps,
        limit: this.config.maxConcurrentOps,
      });
    }

    // Check for recovery
    if (this.previousStatus !== 'healthy' && status === 'healthy') {
      this.emit({
        type: 'resource.recovered',
        metric: 'overall',
        previousStatus: this.previousStatus,
      });
    }

    this.previousStatus = status;

    return { status, usage, canContinue, message, recommendation };
  }

  /**
   * Check if we can start a new operation.
   */
  canStartOperation(): boolean {
    if (!this.config.enabled) return true;
    const check = this.check();
    return check.canContinue && this.concurrentOps < this.config.maxConcurrentOps;
  }

  /**
   * Track the start of an operation.
   * Returns a function to call when the operation completes.
   */
  startOperation(): () => void {
    this.concurrentOps++;
    return () => {
      this.concurrentOps = Math.max(0, this.concurrentOps - 1);
    };
  }

  /**
   * Run a function with operation tracking.
   * Throws ResourceLimitError if resources exceeded.
   */
  async runTracked<T>(fn: () => Promise<T>): Promise<T> {
    const check = this.check();
    if (!check.canContinue) {
      throw new ResourceLimitError(check.message || 'Resource limit exceeded');
    }

    const end = this.startOperation();
    try {
      return await fn();
    } finally {
      end();
    }
  }

  /**
   * Run only if resources are available.
   * Returns fallback value if resources exceeded.
   */
  async runIfAvailable<T>(fn: () => Promise<T>, fallback?: T): Promise<T | undefined> {
    if (!this.canStartOperation()) {
      return fallback;
    }

    const check = this.check();
    if (!check.canContinue || check.status === 'critical') {
      return fallback;
    }

    return this.runTracked(fn);
  }

  /**
   * Update resource limits.
   */
  setLimits(limits: Partial<ResourceLimitsConfig>): void {
    Object.assign(this.config, limits);
  }

  /**
   * Get current limits.
   */
  getLimits(): Required<ResourceLimitsConfig> {
    return { ...this.config };
  }

  /**
   * Reset timing and operation count.
   */
  reset(): void {
    this.startTime = Date.now();
    this.concurrentOps = 0;
    this.previousStatus = 'healthy';
  }

  /**
   * Reset only the CPU time counter (per-prompt reset).
   * This allows long sessions with multiple prompts without hitting the wall-clock limit.
   */
  resetCpuTime(): void {
    this.startTime = Date.now();
  }

  /**
   * Get formatted status string.
   */
  getStatusString(): string {
    const check = this.check();
    const usage = check.usage;

    return [
      `Memory: ${usage.memoryMB.toFixed(1)}MB / ${this.config.maxMemoryMB}MB (${(usage.memoryPercent * 100).toFixed(1)}%)`,
      `CPU Time: ${usage.cpuTimeSec.toFixed(1)}s / ${this.config.maxCpuTimeSec}s (${(usage.cpuPercent * 100).toFixed(1)}%)`,
      `Operations: ${usage.concurrentOps} / ${this.config.maxConcurrentOps}`,
      `Status: ${check.status.toUpperCase()}`,
    ].join('\n');
  }

  /**
   * Subscribe to resource events.
   */
  subscribe(listener: ResourceEventListener): () => void {
    this.eventListeners.add(listener);
    return () => this.eventListeners.delete(listener);
  }

  /**
   * Emit a resource event.
   */
  private emit(event: ResourceEvent): void {
    for (const listener of this.eventListeners) {
      try {
        listener(event);
      } catch {
        // Ignore listener errors
      }
    }
  }

  /**
   * Cleanup resources.
   */
  cleanup(): void {
    this.eventListeners.clear();
  }
}

// =============================================================================
// ERROR
// =============================================================================

/**
 * Error thrown when resource limits are exceeded.
 */
export class ResourceLimitError extends Error {
  readonly isResourceLimit = true;

  constructor(message: string) {
    super(message);
    this.name = 'ResourceLimitError';
  }
}

/**
 * Check if an error is a resource limit error.
 */
export function isResourceLimitError(error: unknown): error is ResourceLimitError {
  return (
    error instanceof ResourceLimitError ||
    (error instanceof Error &&
      'isResourceLimit' in error &&
      (error as ResourceLimitError).isResourceLimit === true)
  );
}

// =============================================================================
// FACTORY FUNCTIONS
// =============================================================================

/**
 * Create a resource manager with default limits.
 */
export function createResourceManager(config?: ResourceLimitsConfig): ResourceManager {
  return new ResourceManager(config);
}

/**
 * Create a strict resource manager (lower limits).
 */
export function createStrictResourceManager(): ResourceManager {
  return new ResourceManager({
    enabled: true,
    maxMemoryMB: 128,
    maxCpuTimeSec: 60,
    maxConcurrentOps: 5,
    warnThreshold: 0.5,
    criticalThreshold: 0.8,
  });
}

/**
 * Create a lenient resource manager (higher limits).
 */
export function createLenientResourceManager(): ResourceManager {
  return new ResourceManager({
    enabled: true,
    maxMemoryMB: 1024,
    maxCpuTimeSec: 1800, // 30 minutes for complex tasks
    maxConcurrentOps: 50,
    warnThreshold: 0.8,
    criticalThreshold: 0.95,
  });
}

// =============================================================================
// INTEGRATION HELPERS
// =============================================================================

/**
 * Combine resource check with economics budget check.
 * Returns true only if BOTH resource AND budget checks pass.
 */
export function combinedShouldContinue(
  resourceManager: ResourceManager | null,
  economicsShouldContinue: boolean,
): { canContinue: boolean; reason?: string } {
  // If no resource manager, just use economics
  if (!resourceManager) {
    return {
      canContinue: economicsShouldContinue,
      reason: economicsShouldContinue ? undefined : 'Budget limit',
    };
  }

  // Check resources
  const resourceCheck = resourceManager.check();

  // Both must pass
  if (!economicsShouldContinue) {
    return { canContinue: false, reason: 'Budget limit' };
  }

  if (!resourceCheck.canContinue) {
    return { canContinue: false, reason: resourceCheck.message || 'Resource limit' };
  }

  return { canContinue: true };
}

File: attocode/attocode_py/src/attocode/integrations/budget/resources.py
"""Resource monitoring and limits.

Tracks CPU time, memory usage, file handles, and concurrent operations
to prevent resource exhaustion during long-running agent sessions.
"""

from __future__ import annotations

import os
import resource
import time
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any


class ResourceStatus(StrEnum):
    """Current resource status."""

    OK = "ok"
    WARNING = "warning"
    CRITICAL = "critical"
    EXHAUSTED = "exhausted"


@dataclass(slots=True)
class ResourceConfig:
    """Resource monitoring configuration."""

    enabled: bool = True
    max_memory_mb: int = 512
    max_cpu_time_sec: int = 1_800  # 30 minutes per prompt
    max_concurrent_ops: int = 10
    warn_threshold: float = 0.7
    critical_threshold: float = 0.9


@dataclass(slots=True)
class ResourceCheck:
    """Result of a resource check."""

    status: ResourceStatus
    memory_usage_mb: float = 0.0
    cpu_time_sec: float = 0.0
    concurrent_ops: int = 0
    message: str = ""
    should_stop: bool = False


@dataclass
class ResourceManager:
    """Monitors and limits system resource usage.

    Tracks CPU time, memory, and concurrent operations. Resets CPU
    tracking per-prompt so limits apply to individual prompt executions.
    """

    config: ResourceConfig = field(default_factory=ResourceConfig)

    _prompt_start_cpu: float = field(default=0.0, repr=False)
    _concurrent_ops: int = field(default=0, repr=False)
    _peak_memory_mb: float = field(default=0.0, repr=False)

    def __post_init__(self) -> None:
        self.reset_prompt()

    def reset_prompt(self) -> None:
        """Reset per-prompt CPU tracking. Call at start of each new prompt."""
        usage = resource.getrusage(resource.RUSAGE_SELF)
        self._prompt_start_cpu = usage.ru_utime + usage.ru_stime

    def acquire_operation(self) -> bool:
        """Try to acquire a slot for a concurrent operation.

        Returns True if allowed, False if at limit.
        """
        if self._concurrent_ops >= self.config.max_concurrent_ops:
            return False
        self._concurrent_ops += 1
        return True

    def release_operation(self) -> None:
        """Release a concurrent operation slot."""
        self._concurrent_ops = max(0, self._concurrent_ops - 1)

    def check(self) -> ResourceCheck:
        """Check current resource usage against limits."""
        if not self.config.enabled:
            return ResourceCheck(status=ResourceStatus.OK)

        # Memory check
        try:
            usage = resource.getrusage(resource.RUSAGE_SELF)
            # ru_maxrss is in bytes on Linux, KB on macOS
            memory_kb = usage.ru_maxrss
            if os.uname().sysname == "Darwin":
                memory_mb = memory_kb / (1024 * 1024)  # bytes -> MB on macOS
            else:
                memory_mb = memory_kb / 1024  # KB -> MB on Linux
        except (OSError, AttributeError):
            memory_mb = 0.0

        self._peak_memory_mb = max(self._peak_memory_mb, memory_mb)

        # CPU time check (per-prompt)
        try:
            usage = resource.getrusage(resource.RUSAGE_SELF)
            cpu_time = (usage.ru_utime + usage.ru_stime) - self._prompt_start_cpu
        except (OSError, AttributeError):
            cpu_time = 0.0

        # Determine status
        status = ResourceStatus.OK
        message = ""
        should_stop = False

        # Memory exhaustion
        if self.config.max_memory_mb > 0:
            mem_ratio = memory_mb / self.config.max_memory_mb
            if mem_ratio >= 1.0:
                status = ResourceStatus.EXHAUSTED
                message = f"Memory limit exceeded: {memory_mb:.0f}MB / {self.config.max_memory_mb}MB"
                should_stop = True
            elif mem_ratio >= self.config.critical_threshold:
                status = ResourceStatus.CRITICAL
                message = f"Memory critical: {memory_mb:.0f}MB / {self.config.max_memory_mb}MB"
            elif mem_ratio >= self.config.warn_threshold:
                status = ResourceStatus.WARNING
                message = f"Memory warning: {memory_mb:.0f}MB / {self.config.max_memory_mb}MB"

        # CPU time exhaustion
        if self.config.max_cpu_time_sec > 0 and not should_stop:
            cpu_ratio = cpu_time / self.config.max_cpu_time_sec
            if cpu_ratio >= 1.0:
                status = ResourceStatus.EXHAUSTED
                message = f"CPU time exceeded: {cpu_time:.0f}s / {self.config.max_cpu_time_sec}s"
                should_stop = True
            elif cpu_ratio >= self.config.critical_threshold and status.value < ResourceStatus.CRITICAL.value:
                status = ResourceStatus.CRITICAL
                message = f"CPU time critical: {cpu_time:.0f}s / {self.config.max_cpu_time_sec}s"

        return ResourceCheck(
            status=status,
            memory_usage_mb=memory_mb,
            cpu_time_sec=cpu_time,
            concurrent_ops=self._concurrent_ops,
            message=message,
            should_stop=should_stop,
        )

    def get_stats(self) -> dict[str, Any]:
        """Get current resource stats."""
        check = self.check()
        return {
            "memory_mb": check.memory_usage_mb,
            "peak_memory_mb": self._peak_memory_mb,
            "cpu_time_sec": check.cpu_time_sec,
            "concurrent_ops": self._concurrent_ops,
            "status": check.status.value,
        }
